### PR TITLE
Change pointer arithmetic to use getelementptr.

### PR DESCRIPTION
**Pull request by rdomingu**
Change inttoptr(add x, const) -> gep(inttoptr x, const) to improve
value tracking and load/store vectorization. Fix regression with
non-array indexing.

### CODE REVIEW

**Comment by rdomingu** [comment]
This is pull request is a follow-up to https://github.com/GPUOpen-Drivers/llpc/pull/1028 which was reverted due to a regression when indexing a non-array structure.

**Comment by JaxLinAMD** [comment]
@rdomingu  please rebase

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 1]
```diff
@@ -571,6 +571,54 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
 
+// =====================================================================================================================
+// Visit an inttoptr instruction.
+//
+// Change inttoptr ( add x, const ) -> gep ( inttoptr x, const ) to improve value tracking and load/store vectorization.
+//
+// Note: we decided to implement this transformation here and not in LLVM. From the point of view of alias analysis, the
+// pointer returned by inttoptr ( add x, const ) is different from the pointer returned by gep ( inttoptr x, const ):
+// the former is associated with whatever x AND const are associated with; the latter is associated ONLY with whatever x
+// is associated with.
```
Comment (kuhar) [inline lgc/patch/PatchPeepholeOpt.cpp:582]: I'd reword it like this:
```
the former is associated with whatever x AND const point to; the latter is associated ONLY with whatever x points to
```

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 2]
```diff
@@ -571,6 +571,54 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
+//
+// We can assume in LLPC that const is not associated with any object (which makes this transformation valid) but that's
```
Comment (kuhar) [inline lgc/patch/PatchPeepholeOpt.cpp:584]: I'd reword this like this:
```
In LLPC/LGC, we can assume that const does not point to any object (which makes this transformation valid) but that's 
```

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 3]
```diff
@@ -571,6 +571,54 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
+    return;
+
+  // Create a getelementptr instruction (using offset / size).
+  const DataLayout dataLayout = intToPtr.getModule()->getDataLayout();
+  const uint64_t size = dataLayout.getTypeAllocSize(intToPtr.getType()->getPointerElementType());
+  APInt index = constOffset->getValue().udiv(size);
+  if (constOffset->getValue().urem(size) != 0)
+    return;
+
+  // Change inttoptr ( add x, const ) -> gep ( inttoptr x, const )
```
Comment (kuhar) [inline lgc/patch/PatchPeepholeOpt.cpp:606]: nit: missing comma at the end

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 4]
```diff
@@ -571,6 +571,54 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
+// the former is associated with whatever x AND const are associated with; the latter is associated ONLY with whatever x
+// is associated with.
+//
+// We can assume in LLPC that const is not associated with any object (which makes this transformation valid) but that's
+// not an assumption that can be made in general in LLVM with all its different front-ends.
+//
+// @param intToPtr: The "inttoptr" instruction to visit.
+void PatchPeepholeOpt::visitIntToPtr(IntToPtrInst &intToPtr) {
+  // Check if we are using add to do pointer arithmetic.
+  BinaryOperator *const binaryOperator = dyn_cast<BinaryOperator>(intToPtr.getOperand(0));
```
Comment (kuhar) [inline lgc/patch/PatchPeepholeOpt.cpp:590]: nit: use auto, the type is clear from the right hand side

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 5]
```diff
@@ -571,6 +571,54 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
+  if (!binaryOperator || binaryOperator->getOpcode() != Instruction::Add)
+    return;
+
+  // Check that we have a constant offset.
+  const ConstantInt *const constOffset = dyn_cast<ConstantInt>(binaryOperator->getOperand(1));
```
Comment (kuhar) [inline lgc/patch/PatchPeepholeOpt.cpp:595]: nit: use auto, the type is clear from the right hand side

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 6]
```diff
@@ -571,6 +571,54 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
+  if (!constOffset)
+    return;
+
+  // Create a getelementptr instruction (using offset / size).
+  const DataLayout dataLayout = intToPtr.getModule()->getDataLayout();
```
Comment (kuhar) [inline lgc/patch/PatchPeepholeOpt.cpp:600]: missing reference?

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 7]
```diff
@@ -571,6 +571,54 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
+  const uint64_t size = dataLayout.getTypeAllocSize(intToPtr.getType()->getPointerElementType());
+  APInt index = constOffset->getValue().udiv(size);
+  if (constOffset->getValue().urem(size) != 0)
+    return;
+
+  // Change inttoptr ( add x, const ) -> gep ( inttoptr x, const )
+  IntToPtrInst *const newIntToPtr = new IntToPtrInst(binaryOperator->getOperand(0), intToPtr.getType());
```
Comment (kuhar) [inline lgc/patch/PatchPeepholeOpt.cpp:607]: nit: use auto, the type is clear from the right hand side

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 8]
```diff
@@ -571,6 +571,54 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
+  insertAfter(*newIntToPtr, *binaryOperator);
+
+  GetElementPtrInst *const getElementPtr =
```
Comment (kuhar) [inline lgc/patch/PatchPeepholeOpt.cpp:610]: nit: use auto, the type is clear from the right hand side

**Review by kuhar — CHANGES_REQUESTED** [review]

File: lgc/patch/PatchPeepholeOpt.cpp (hunk 2 above)
Comment (kuhar) [reply lgc/patch/PatchPeepholeOpt.cpp:584]: Can you also paste the link to the llvm-dev discussion from the other PR?

File: lgc/patch/PatchPeepholeOpt.cpp (hunk 3 above)
Comment (rdomingu) [reply lgc/patch/PatchPeepholeOpt.cpp:606]: Where?

File: lgc/patch/PatchPeepholeOpt.cpp (hunk 6 above)
Comment (rdomingu) [reply lgc/patch/PatchPeepholeOpt.cpp:600]: Where?

Comment (rdomingu) [reply lgc/patch/PatchPeepholeOpt.cpp:600]: Nevermind. I got it.

File: lgc/patch/PatchPeepholeOpt.cpp (hunk 3 above)
Comment (rdomingu) [reply lgc/patch/PatchPeepholeOpt.cpp:606]: You mean missing _period_ at the end?

Comment (kuhar) [reply lgc/patch/PatchPeepholeOpt.cpp:606]: ah, yes

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1294ed0f61226916b9ed4ac40ca2a86d2970981c
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 8abed125acda77ccf700ddd0f5bf0114d25ac135</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 2f28502c5e6e4f059a014ab30001a191328f7a5d</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 0472e039337a255a85b0f6f4ac927aa19261c2f0</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/192520)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 26332/48178 (54.7%)</li>
<li>Failed: 0/48178 (0.0%)</li>
<li>Skipped: 21846/48178 (45.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26331/48178 (54.7%)</li>
<li>Failed: 0/48178 (0.0%)</li>
<li>Skipped: 21847/48178 (45.3%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 24863/48157 (51.6%)</li>
<li>Failed: 0/48157 (0.0%)</li>
<li>Skipped: 23294/48157 (48.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 26230/48007 (54.6%)</li>
<li>Failed: 0/48007 (0.0%)</li>
<li>Skipped: 21777/48007 (45.4%)</li>
</ul></details>


</ul></details>

**Review by kuhar — APPROVED** [review]
Thanks for the fixes

**Review by amdrexu — APPROVED** [review]
@rdomingu Please help to click the button "Close Conversation" once you fix a review issue pointed by review comments in the future. This could help the one who made the comments to quickly check your update as a reminder. Thanks very much.

**Comment by JaxLinAMD** [comment]
@rdomingu  ping

**Comment by rdomingu** [comment]
> @rdomingu Please help to click the button "Close Conversation" once you fix a review issue pointed by review comments in the future. This could help the one who made the comments to quickly check your update as a reminder. Thanks very much.

Will do. Thank you.

**Comment by rdomingu** [comment]
> @rdomingu ping

I am not sure what this ping is for.

**Comment by JaxLinAMD** [comment]
> > @rdomingu ping
> 
> I am not sure what this ping is for.

the same thing "Close conversion" :)
